### ISSUE TITLE
BREAKING: Port to OpenTelemetry

### ISSUE THREAD

**Issue by jefflill** [nforgeio/neonKUBE#1662]
I've spent a bit of time skimming over the OpenTelemetry docs, wondering if now might be the time to switch over from LogManager and Prometheus.

I've always wondered how tracing actually works.  In the past, the presentations were somewhat unclear about how trace data was actually exported in common situations and new presentations are still somewhat unclear (at least to me).

For the Kubernetes world, it seems like our services would need to configure a trace exporter that communicates with Tempo by sending trace information.  This means that NeonService or something would need to know the internal service name and port for Tempo and be able to discover this somehow automatically or have this info passed into the app container as environment variables so we can configure the exporter.

Extending this to apps like neon-desktop or the Loopie phone apps implies that we'd need two exporters running in the app: one for exporting traces to the cluster and the other exporting logs.  Presumably, we'd need headend services that authenticate the app/user and then forward this data to Tempo and Loki.
It looks like we'd need to do the same thing for metrics, where we'd probably need to write a custom metrics exporter that we'd include in the app to forward metrics to a headend API which forwards to Mimir.

For extra credit, we might make these headend services configurable so they could forward data to other places like Grafana Cloud, etc.

This seems like it could be a fairly deep rabbit hole that I'm not entirely sure we want to go down right now.  On the other hand, we're effectively talking about writing a LogManager exporter for neon-desktop, so rather than doing that we could just write an OpenTelemetry exporter, perhaps getting some help from existing opensource exporters.

Regardless, we'd also need to figure out something for trace exporting.  We might be able to ditch this temporarily and rely solely on the trace IDs in logs but we'd lose trace span baggage which means that we'd have to add a bunch of log labels like you did for Loopie which is messy and will lead to code/log bloat.  I don't think we really want to standardize on that approach.

I'm a bit torn about this:

* It would be nice to hop on this bandwagon and ditch the ancient LogManager code (I probably coded the original version way back in 2005).

* This would probably need to be a breaking change since this is used all over the Neon libraries.  We'd probably want to fork a v2.x branch for the older version of the libraries and bump to v3.x for the next gen libs.

* This might also be a good time to split the neonkube and neonsdk repos although I don't think that would be strictly necessary.
I could see us implementing the log, metrics, and trace exporters for client apps in neonkube and perhaps the cluster side receivers as neonkube services that our users could use as well, so our headend services would just deploy these services, probably behind and application gateway that handles authentication and throttling.

* This will be a substantial change for the neon libraries but probably not too terrible for our or Loopie's service apps, since we're loosly based on MSFT's logging extensions.

* It would also be nice to switch to OpenTelemetry before we author a bunch of new services for the headend as well as services we might add to neonkube.  Now that neonkube 1.0 is getting close, we're poised to begin writing probably a substantial number of new apps soon.

* I can see writing _"exporters"_ for our existing logger taking a few days and we'd also need to implement the headend side as well.  We'd need to do something similar for tracing, although I hope we could take advantage of existing exporters for tracing.

* Finally, I'm not entirely comfortable with doing YouTube explainers using our existing LogManager and NeonService implementations since those are somewhat non-standard now and if we deferred this change for a few months, early explainers will be out-of-date and we might end up leading adopters astray.

I'm thinking about having a go at this now, first by splitting the neonkube/neonsdk repos and then forking a main-v2 branch in neonsdk to support people using the older code (mostly Loopie, the new RaspberryDebugger guy, and neonkube until the port is completed) and then convert to OpenTelemetry.

We should discuss this.

**Comment by jefflill** [comment]
I've spent a bit of time skimming over the OpenTelemetry docs, wondering if now might be the time to switch over from LogManager and Prometheus.

I've always wondered how tracing actually works.  In the past, the presentations were somewhat unclear about how trace data was actually exported in common situations and new presentations are still somewhat unclear (at least to me).

For the Kubernetes world, it seems like our services would need to configure a trace exporter that communicates with Tempo by sending trace information.  This means that NeonService or something would need to know the internal service name and port for Tempo and be able to discover this somehow automatically or have this info passed into the app container as environment variables so we can configure the exporter.

Extending this to apps like neon-desktop or the Loopie phone apps implies that we'd need two exporters running in the app: one for exporting traces to the cluster and the other exporting logs.  Presumably, we'd need headend services that authenticate the app/user and then forward this data to Tempo and Loki.
It looks like we'd need to do the same thing for metrics, where we'd probably need to write a custom metrics exporter that we'd include in the app to forward metrics to a headend API which forwards to Mimir.

For extra credit, we might make these headend services configurable so they could forward data to other places like Grafana Cloud, etc.

This seems like it could be a fairly deep rabbit hole that I'm not entirely sure we want to go down right now.  On the other hand, we're effectively talking about writing a LogManager exporter for neon-desktop, so rather than doing that we could just write an OpenTelemetry exporter, perhaps getting some help from existing opensource exporters.

Regardless, we'd also need to figure out something for trace exporting.  We might be able to ditch this temporarily and rely solely on the trace IDs in logs but we'd lose trace span baggage which means that we'd have to add a bunch of log labels like you did for Loopie which is messy and will lead to code/log bloat.  I don't think we really want to standardize on that approach.

I'm a bit torn about this:

* It would be nice to hop on this bandwagon and ditch the ancient LogManager code (I probably coded the original version way back in 2005).

* This would probably need to be a breaking change since this is used all over the Neon libraries.  We'd probably want to fork a v2.x branch for the older version of the libraries and bump to v3.x for the next gen libs.

* This might also be a good time to split the neonkube and neonsdk repos although I don't think that would be strictly necessary.
I could see us implementing the log, metrics, and trace exporters for client apps in neonkube and perhaps the cluster side receivers as neonkube services that our users could use as well, so our headend services would just deploy these services, probably behind and application gateway that handles authentication and throttling.

* This will be a substantial change for the neon libraries but probably not too terrible for our or Loopie's service apps, since we're loosly based on MSFT's logging extensions.

* It would also be nice to switch to OpenTelemetry before we author a bunch of new services for the headend as well as services we might add to neonkube.  Now that neonkube 1.0 is getting close, we're poised to begin writing probably a substantial number of new apps soon.

* I can see writing _"exporters"_ for our existing logger taking a few days and we'd also need to implement the headend side as well.  We'd need to do something similar for tracing, although I hope we could take advantage of existing exporters for tracing.

* Finally, I'm not entirely comfortable with doing YouTube explainers using our existing LogManager and NeonService implementations since those are somewhat non-standard now and if we deferred this change for a few months, early explainers will be out-of-date and we might end up leading adopters astray.

I'm thinking about having a go at this now, first by splitting the neonkube/neonsdk repos and then forking a main-v2 branch in neonsdk to support people using the older code (mostly Loopie, the new RaspberryDebugger guy, and neonkube until the port is completed) and then convert to OpenTelemetry.

We should discuss this.

**Comment by jefflill** [comment]
This is going to be a breaking change for neonSDK v3.0+.

- [ ] Remove `ServiceContainer` and replace it with the standard Microsoft `ServiceCollection` class?

    So our `ServiceContainer` class is an enhancement over MSFT's `ServiceCollection`.  The benefit is that `ServiceContainer` allows for service lookup at anytime whereas `ServiceCollection` allows lookup only after `IServiceCollection.BuildServiceProvider()` is called.  

After thinking about it, I think that the reason MSFT went with this approach is so that the `ServiceCollection` will effectively be read-only after the build call and won't require any locking thereafter.  We need to use a lock in our implementation which could easily result in a performance hit for things like high traffic websites where thinks may need to be injected into ASP.NET controller constructors.

I think I can tweak `ServiceContainer` to get make the think read-only after the `Build` call and not use a lock thereafter.

- [ ] Can we somehow unify the global `IServiceCollection` for both ASP.NET and Console apps so that `NeonService` can work the same for both environments?

- [ ]

**Comment by jefflill** [comment]
Manually transferred this issue to neonSDK.
